hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/InsertAndSelectParser.java

```diff
@@ -0,0 +1,78 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.persister.entity.impl;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.hibernate.dialect.CockroachDB192Dialect;
+import org.hibernate.dialect.DB2Dialect;
+import org.hibernate.dialect.Dialect;
+import org.hibernate.dialect.PostgreSQL81Dialect;
+import org.hibernate.dialect.SQLServerDialect;
+
+/**
+ * Utility methods to parse SQL and create the string to get the generated id value during an insert.
+ * <p>
+ *     This is an issue with Hibernate ORM 5 and should be unnecessary with Hibernate ORM 6.
+ * <p>
+ */
+// TODO: Add some details about the issue in ORM 5 that prevents us to generate the queries in advance
+public class InsertAndSelectParser {
+
+	private InsertAndSelectParser() {
+	}
+
+	/**
+	 * Matches a SQL server query of this type ({@code default} is optional):
+	 * <pre>
+	 *     insert into entity_table default values select id, name from other_table;
+	 * </pre>
+	 */
+	private static final Pattern MSSQL_PATTERN = Pattern.compile( "^(insert into .*?)(( default)? values.*)" );
```

Comment:
Can we please not use a regex for this?

_Especially_ not a regex which matches `.*`.

Reply:
This should solve the issue: https://github.com/hibernate/hibernate-reactive/pull/865/commits/196658305fc17a2f413c6345f22fc6a61c0da3bb

It still uses a regex but only to check the last part of the string end now queries are generated once.

Reply:
What exactly is that regex supposed to be doing? Matching parentheses? **And** assuming that you never have more than one level of nesting?

Reply:
It matches the pattern `values (?, ?, ?, ?, ?)` in a query of type `insert into Table (col1, col2) values (?,?)`. Where the number of parameters varies. We only use it for query that have to return the id after an insert. What kind of nesting can we have?

Reply:
But I guess this is not going to work if somebody uses a `@ColumnTransformer`. Uff, I will add a test tomorrow

Reply:
I think regexes are just terrible in at almost everything, but they're particularly terrible at parsing programming languages, since they simply can't handle any language with a recursive grammar (which is every language).

And look, _perhaps_ there's no scenario right now where we let you have an expression in an `insert` list, but SQL certainly allows it, so it's just not robust. I already ripped out the use of broken regexes from some of the dialects in H6, and used handwritten parsers instead. The results was less code, and less-broken code. And it's easier to understand.

Reply:
This approach should work in all cases: https://github.com/hibernate/hibernate-reactive/pull/865/commits/80f7691e06464cad18fc2e35a708783d371344e3

Although, it feels like we should send a patch to ORM but I haven't looked too close on why the queries are different. For now I was interested in finding a solution for Reactive.